Sources/NIOEmbedded/AsyncTestingChannel.swift

```diff
@@ -0,0 +1,521 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+import NIOConcurrencyHelpers
+import NIOCore
+
+/// `NIOAsyncTestingChannel` is a `Channel` implementation that does no
+/// actual IO but that does have proper eventing mechanism, albeit one that users can
+/// control. The prime use-case for `NIOAsyncTestingChannel` is in unit tests when you
+/// want to feed the inbound events and check the outbound events manually.
+///
+/// Please remember to call `finish()` when you are no longer using this
+/// `NIOAsyncTestingChannel`.
+///
+/// To feed events through an `NIOAsyncTestingChannel`'s `ChannelPipeline` use
+/// `NIOAsyncTestingChannel.writeInbound` which accepts data of any type. It will then
+/// forward that data through the `ChannelPipeline` and the subsequent
+/// `ChannelInboundHandler` will receive it through the usual `channelRead`
+/// event. The user is responsible for making sure the first
+/// `ChannelInboundHandler` expects data of that type.
+///
+/// `NIOAsyncTestingChannel` automatically collects arriving outbound data and makes it
+/// available one-by-one through `readOutbound`.
+///
+/// - note: `NIOAsyncTestingChannel` is currently only compatible with
+///   `NIOAsyncTestingEventLoop`s and cannot be used with `SelectableEventLoop`s from
+///   for example `MultiThreadedEventLoopGroup`.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
```

Comment:
Maybe we can add a little diagram here to visualise the flow of data, just a quick draw:
```suggestion
///
///                            readOutbound
///                                 │
///                                 │
///                                 ▼
///                          ┌──────────────┐
///            writeInbound  │OutboundBuffer│
///                          └──────────────┘
///                 │               ▲
///                 │               │
///                 ▼               │
///             ┌────────────────────────┐
///             │ NIOAsyncTestingChannel │
///             └────────────────────────┘
///                 │                ▲
///                 │                │
/// fireChannelRead │                │  write & flush
///                 │                │
///                 ▼                │
///             ┌────────────────────────┐
///             │     ChannelHandler     │
///             └────────────────────────┘
///                 │                ▲
/// fireChannelRead │                │
///                 ▼                │
///           ┌─────────────┐
///           │InboundBuffer│  writeOutbound
///           └─────────────┘
///                 ▲
///                 │
///                 │
///            readInbound
///
@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
```